---
title: "probe_based_genotyping_apoe_adjunct"
output: html_document
date: "2025-10-30"
---

This script is to be run as and adjunct analysis after successfully completing probe_based_genotyping.Rmd. Resulting output from probe_based_genotyping.Rmd is imported for additional workup and analysis.

# 1. Setup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)

root_dir_path <- readline(prompt = "Please enter the root directory path (e.g., /Users/user/Box/project/): ")
knitr::opts_knit$set(root.dir = root_dir_path)
root.dir <- getwd()
```

# 2. Data import and preparation
```{r}
ratio_apoe_data <- readline(prompt = "Please apoe ratio data path (e.g., data/apoe_ratio_data.csv): ") 
ratio_apoe_data <- read_csv(ratio_apoe_data)# Warning may display due to NA or missing data in some columns
ratio_apoe_data <- ratio_apoe_data[,-1] # Remove unnecessary column generated by import

ratio_apoe_data
```
# 3. Assign APOE ID

## a. Create APOE ID reference
```{r}
Allele_col_ID <- c("e2/e2", "e2/e3", "e3/e3", "e2/e4", "e3/e4", "e4/e4") #APOE ID per SNP pair
rs429358 <- c("T/T", "T/T", "T/T", "C/T", "C/T", "C/C") #rs429358 allele combination
rs7412 <- c("T/T", "C/T", "C/C", "C/T", "C/C", "C/C") #rs7412 allele combination

# Create the data frame
apoe_genotype_reference <- data.frame(Allele_ID = Allele_col_ID, rs429358 = rs429358, rs7412 = rs7412) #Create data frame

apoe_genotype_reference$apoe_allele <- paste(apoe_genotype_reference$rs429358, apoe_genotype_reference$rs7412) #Combine columns for efficient APOE ID per Alternate_ID
apoe_genotype_reference
```
## b. Format dataset for APOE ID assignment
```{r}
#Format already ordered genotypes for APOE ID according to reference
ratio_apoe_data$apoe_allele <- paste(ratio_apoe_data$rs429358, ratio_apoe_data$rs7412) #Combine columns to match reference
ratio_apoe_data = select(ratio_apoe_data, c("Alternate_ID", "apoe_allele")) #Simplify data frame

ratio_apoe_data
```

## c. Assign APOE ID
```{r}
# Add APOE genotype labels to each sample by merging with the reference table
ratio_apoe_data <- merge(ratio_apoe_data, apoe_genotype_reference, # merge tables `ratio_apoe_data` and `apoe_genotype_reference`, combining by common column `apoe_allele`
                         by = "apoe_allele", # column in ratio_data used for merging, effectively assigning APOE ID bases via `Allele_ID` column in `apoe_genotype_reference`
                         all.x = TRUE) # Keep all rows in `ratio_apoe_data` even if no match exists in `apoe_genotype_reference`

ratio_apoe_data = select(ratio_apoe_data, c("Alternate_ID","rs429358", "rs7412", "Allele_ID"))

write.csv(ratio_apoe_data, "R_output/ratio_apoe_data.csv")

ratio_apoe_data
```